transport/http/checksum_middleware_test.go

```diff
@@ -0,0 +1,80 @@
+package http
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"strings"
+	"testing"
+
+	smithyio "github.com/awslabs/smithy-go/io"
+	"github.com/awslabs/smithy-go/middleware"
+)
+
+func TestChecksumMiddleware(t *testing.T) {
+	cases := map[string]struct {
+		payload               io.ReadSeeker
+		expectedPayloadLength int64
+		expectedMD5Checksum   string
+		expectError           string
+	}{
+		"empty body": {
+			payload: smithyio.ReadSeekNopCloser{
+				ReadSeeker: bytes.NewReader([]byte(``)),
+			},
+			expectedPayloadLength: 0,
+			expectedMD5Checksum:   "1B2M2Y8AsgTpgAmY7PhCfg==",
+		},
+		"standard req body": {
+			payload: smithyio.ReadSeekNopCloser{
+				ReadSeeker: bytes.NewReader([]byte(`abc`)),
+			},
+			expectedPayloadLength: 3,
+			expectedMD5Checksum:   "kAFQmDzST7DWlj99KOF/cg==",
+		},
+		"nil body": {},
```

Comment:
Would be helpful to have a payload that is a `bytes.Buffer` without `ReadSeekNopCloser` to exercise the `RewindStream` method failing because the underlying buffer isn't seekable.